### PR TITLE
QBMapper id doesn't have to be an int

### PR DESCRIPTION
**Pull request by rullzer**
Found when looking at https://github.com/nextcloud/mail/issues/1969

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

### CODE REVIEW

**Comment by kesselb** [comment]
https://github.com/nextcloud/server/blob/5aeb8eac2bcfd13760d779aec9027ffc955011d9/lib/public/AppFramework/Db/QBMapper.php#L123-L125

If someone uses a string as id the value is already set before insert. So id is not null and getLastInsertId() is not called.

**Comment by rullzer** [comment]
Ah mmm right... meh. let me think some moreabout it then.

**Comment by kesselb** [comment]
Sorry :see_no_evil: Let me rephrase it: I don't think we have to care about the case because its unlikely. If someone use a hash as id the hash should be set before the insert. We could throw a exception if the type is not integer and the id is null.

File: lib/public/AppFramework/Db/QBMapper.php [hunk 1]
```diff
@@ -187,7 +187,7 @@ public function update(Entity $entity): Entity {
 		}
 
 		$qb->where(
-			$qb->expr()->eq('id', $qb->createNamedParameter($id, IQueryBuilder::PARAM_INT))
+			$qb->expr()->eq('id', $qb->createNamedParameter($id, $this->getParameterTypeForProperty($entity, 'id')))
```
Comment (ChristophWurst) [inline lib/public/AppFramework/Db/QBMapper.php]: This was also my consideration. The only problem with this is the default type. It should be a new optional parameter with `string` as default and overwritten in this case

Comment (ChristophWurst) [reply lib/public/AppFramework/Db/QBMapper.php]: oh, `id` has a default type of `integer` set. so this should actually work then

**Comment by ChristophWurst** [comment]
> I don't think we have to care about the case because its unlikely. If someone use a hash as id the hash should be set before the insert. We could throw a exception if the type is not integer and the id is null.

See mail. This happens when you try to run an *update* on an existing row. Nextcloud will just take the `id` as int, even if it's a string. Then the generated query results in a SQL error.

*insert*s actually work as for every column the custom type is respected.

**Comment by kesselb** [comment]
Sorry for the confusion :see_no_evil: 

https://github.com/nextcloud/server/blob/5aeb8eac2bcfd13760d779aec9027ffc955011d9/lib/public/AppFramework/Db/QBMapper.php#L123-L125

For entities without id the lastInsertId is set as id. A entity with `/** @var string $id **/` and no value will :boom: (if the setId method is properly typed of course).

> I don't think we have to care about the case because its unlikely. If someone use a hash as id the hash should be set before the insert. We could throw a exception if the type is not integer and the id is null.

I was wondering if we need to care about this edge case. If someone complains we can still fix it later. I think for the mail app you usually try to insert entities with ids.

**Comment by ChristophWurst** [comment]
Yep, lastInsertId is indeed a problem the way it's used here. 

> I was wondering if we need to care about this edge case. If someone complains we can still fix it later. I think for the mail app you usually try to insert entities with ids.

So the problem is actually inserting a row without an `id` value set and not having an auto increment `id` column. You can also have an `int` `id` without auto increment (for something where the `id` is provided and not auto assigned) and it will do :boom:

**Comment by kesselb** [comment]
> So the problem is actually inserting a row without an `id` value set and not having an auto increment `id` column. You can also have an `int` `id` without auto increment (for something where the `id` is provided and not auto assigned) and it will do boom

You may interested in this former discussion: https://github.com/nextcloud/server/pull/13086

**Review by kesselb — APPROVED** [review]

**Review by ChristophWurst — APPROVED** [review]

**Comment by rullzer** [comment]
Ok maybe this isn't the most elegant way... let me consider a different approach later this evening.
